src/vos/vos_obj.c

```diff
@@ -282,10 +287,18 @@ vos_obj_punch(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch,
 	if (rc == 0 && read_conflict)
 		rc = -DER_TX_RESTART;
 
-	if (dth != NULL && rc == 0)
+	if (rc == 0 && dth != NULL && dth->dth_last_modification)
```

Comment:
this feels a little hacky to do the dth_last_modification thing but the alternative is to split out begin/end into an API and that has issues too

Reply:
Hm, we already have the dtx_being()/dtx_end() APIs that will be used by object layer users. If we introduce another vos_dtx_begin() and vos_dtx_end(), it may be some confused. Another possible solution is that: we use dtx_handle::dth_op_seq to pass the minor epoch for VOS, if we pass another parameter "op_max" via the DTX handle, then VOS can know whether it is the last modification by comparing them. I am not sure which one is better.